builder/rust/cargo/project_aspect.bzl

```diff
@@ -0,0 +1,266 @@
+    for dep in crate_info.deps:
+        if CargoProjectInfo in dep:
+            dep_info = dep[CrateInfo]
+            project_info = dep[CargoProjectInfo]
+            dep_manifest = ctx.actions.declare_file(workspace_root + "/" + dep_info.crate_name + "/" + project_info.manifest.basename)
+            workspace_files.append(dep_manifest)
+            _copy_to_bin(ctx, project_info.manifest, dep_manifest)
+            for path, file in project_info.sources.items():
+                dst = ctx.actions.declare_file(workspace_root + "/" + dep_info.crate_name + "/" + path)
+                _copy_to_bin(ctx, file, dst)
+                workspace_files.append(dst)
+
+    return CargoProjectInfo(
+        manifest = manifest_file,
+        sources = project_sources,
+        workspace_files = workspace_files,
+    )
+
+def _crate_deps(ctx, target):
+    return [dep for dep in _all_deps(ctx) if _TARGET_TYPES[dep[CrateInfo].kind] in ["bin", "lib"]]
+
+def _crate_build_deps(ctx, target):
+    return [dep for dep in _deps(ctx) if _TARGET_TYPES[dep[CrateInfo].kind] == "build"]
+
+def _deps(ctx):
+    return [dep for dep in getattr(ctx.rule.attr, "deps", []) if CrateInfo in dep]
+
+def _all_deps(ctx):
+    return [
+        dep for dep in
+        _deps(ctx) + getattr(ctx.rule.attr, "proc_macro_deps", []) + ([ctx.rule.attr.crate] if hasattr(ctx.rule.attr, "crate") and ctx.rule.attr.crate else [])
+        if CrateInfo in dep
+    ]
+
+def _copy_to_bin(ctx, src, dst):
+    ctx.actions.run_shell(
+        inputs = [src],
+        outputs = [dst],
+        command = "cp -f '{}' '{}'".format(src.path, dst.path),
+    )
+
+def _should_generate_cargo_project(ctx, target):
+    return (str(target.label).startswith("@vaticle") or str(target.label).startswith("//")) and \
+        ctx.rule.kind in _TARGET_TYPES and _TARGET_TYPES[ctx.rule.kind] in ["bin", "lib", "test"]
+
+def _is_universe_crate(target):
+    return str(target.label).startswith("@crates__")
+
+def _build_cargo_properties_file(target, ctx, source_files, crate_info):
+    properties_file = ctx.actions.declare_file("{}.cargo.properties".format(ctx.rule.attr.name))
+    properties = _get_properties(target, ctx, source_files, crate_info)
+
+    content = ""
+    for prop in properties.items():
+        content = content + ("{}: {}\n".format(prop[0], prop[1]))
+    ctx.actions.write(
+        output = properties_file,
+        content = content
+    )
+    return properties_file
+
+def _get_properties(target, ctx, source_files, crate_info):
+    target_type = "build" if _looks_like_cargo_build_script(target) else _TARGET_TYPES[ctx.rule.kind]
+
+    properties = {}
+    properties["name"] = crate_info.crate_name
+    properties["target.name"] = target.label.name
+    properties["type"] = target_type
+    properties["version"] = crate_info.version
+    if target_type in ["bin", "lib"]:
+        properties["edition"] = ctx.rule.attr.edition or "2021"
```

Comment:
we can parametrise these constants like `2021` to some constants at the top of the file